src/tools/wasm-split.cpp

```diff
@@ -245,8 +261,195 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
+  writeProfile->name = Names::getValidFunctionName(*wasm, profileExport);
+  writeProfile->hasExplicitName = true;
+  writeProfile->sig = Signature({Type::i32, Type::i32}, Type::i32);
+  writeProfile->setLocalName(0, "addr");
+  writeProfile->setLocalName(1, "size");
+
+  // Calculate the size of the profile:
+  //   8 bytes module hash +
+  //   4 bytes for the timestamp for each function
+  size_t numDefinedFunctions = 0;
+  ModuleUtils::iterDefinedFunctions(*wasm,
+                                    [&](Function*) { numDefinedFunctions++; });
+  const size_t profileSize = 8 + 4 * numDefinedFunctions;
+
+  // Create the function body
+  Builder builder(*wasm);
+  auto getAddr = [&]() { return builder.makeLocalGet(0, Type::i32); };
+  auto getSize = [&]() { return builder.makeLocalGet(1, Type::i32); };
+  auto hashConst = [&]() { return builder.makeConst(int64_t(moduleHash)); };
+  auto profileSizeConst = [&]() {
+    return builder.makeConst(int32_t(profileSize));
+  };
+
+  // Write the hash followed by all the time stamps
+  Expression* writeData =
+    builder.makeStore(8, 0, 1, getAddr(), hashConst(), Type::i64);
+
+  uint32_t offset = 8;
+  for (const auto& global : functionGlobals) {
+    writeData = builder.blockify(
+      writeData,
+      builder.makeStore(4,
+                        offset,
+                        1,
+                        getAddr(),
+                        builder.makeGlobalGet(global, Type::i32),
+                        Type::i32));
+    offset += 4;
+  }
+
+  writeProfile->body = builder.blockify(
+    builder.makeIf(builder.makeBinary(GeUInt32, getSize(), profileSizeConst()),
+                   writeData),
+    profileSizeConst());
+
+  // Create an export for the function
+  auto ex = std::make_unique<Export>();
+  ex->name = profileExport;
+  ex->value = writeProfile->name;
+  ex->kind = ExternalKind::Function;
+
+  wasm->addFunction(std::move(writeProfile));
+  wasm->addExport(std::move(ex));
+
+  // Also make sure there is a memory with enough pages to write into
+  size_t pages = (profileSize + Memory::kPageSize - 1) / Memory::kPageSize;
+  if (!wasm->memory.exists) {
+    wasm->memory.exists = true;
+    wasm->memory.initial = pages;
+    wasm->memory.max = pages;
+  } else if (wasm->memory.initial < pages) {
+    wasm->memory.initial = pages;
+    if (wasm->memory.max < pages) {
+      wasm->memory.max = pages;
+    }
+  }
+}
+
 void instrumentModule(Module& wasm, const WasmSplitOptions& options) {
-  Fatal() << "TODO: implement instrumentation\n";
+  // Check that the profile export name is not already taken
+  if (wasm.getExportOrNull(options.profileExport) != nullptr) {
+    Fatal() << "error: Export " << options.profileExport << " already exists.";
+  }
+
+  // TODO: calculate module hash.
+  uint64_t moduleHash = 0;
+  PassRunner runner(&wasm, options.passOptions);
+  runner.add(
+    std::make_unique<FuncInstrumentor>(options.profileExport, moduleHash));
+  runner.run();
```

Comment:
I think you can replace these 3 lines with `FuncInstrumentor(..).run(runner, module);`